### ISSUE TITLE
Random Exception (9) from Web Server

### ISSUE THREAD

**Issue by psYbR** [esp8266/Arduino#869]
I frequently get this exception and subsequent reset of the ESP8266: http://i.imgur.com/gCp89u2.png

My code (http://pastebin.com/g4T2Bi8u) is essentially just receiving HTTP requests and outputting the value to serial. The client (same .ino file, flashed to another ESP8266, just change SERVER to false) is another ESP8266 spamming HTTP requests at the server. The client is stable.

It can run for 5 minutes or only a few seconds before throwing the exception, it isn't consistent.

I am not running the latest staging however @luc-github is and can replicate the issue as per his comment on issue #428. I'll update to the latest staging and report any changes here

It may be worth mentioning on a couple of occasions I've seen different exceptions but as I'm not running the latest staging they may already be fixed (and I haven't been able to get them to occur again)

Cheers

**Comment by psYbR** [comment]
Hi @igrr have you had a chance to have a look at this one? I am still getting this quite often

**Comment by igrr** [comment]
I did take a look, but I wasn't able to reproduce as I don't have two ESPs with me currently. 
I've used your SERVER part and sent requests using netcat. Wasn't able to get it to crash, although it does run out of memory quickly. The reason for this is that your server side is doing active close of TCP connection, so the connection remains in TIME_WAIT state for two minutes per TCP spec.

**Comment by psYbR** [comment]
ah I see. thanks @igrr! I'll alter the code not to close the connection each time and see if that fixes the crashing.

**Comment by igrr** [comment]
In the case where you have ESP8266 on both sides (client and server), there isn't much you can do — one side will still have to do active close.
If you have ESP8266 on one side (server only), then you need to make sure client does active close. When using HTTP, there is "Connection: close" header that you may send.

Yet another option is to abort the connection by sending RST segment. There is `client.abort()` method for that. This avoids entering TIME_WAIT state on both sides, but is somewhat a hacky approach.

**Comment by psYbR** [comment]
Thanks for the heads up :) Diverting a little away from ESP8266 questions and more towards HTTP in general I guess but is it possible to perform client.connect() once and then do client.print() each loop? Or do I need to perform a client.close() after each print? I am wanting to control a robot in realtime so the stream of data needs to be continuous and as fast as possible. Is this the best way to do it?

cheers

**Comment by andydarwin** [comment]
I got Similar Problem... I am making a program that can control LED through WIFI by combining Relay Control example from aRest.h 
and Live IP Camera with it.. so i can see the LED turned on or off directly

I frequently got This ERROR... Can anybody tell me Whats going on HERE ????

WiFi connected
Server started
192.168.1.222

Exception (9):
epc1=0x402153c6 epc2=0x00000000 epc3=0x00000000 excvaddr=0x40214cc9 depc=0x00000000

ctx: sys 
sp: 3ffffde0 end: 3fffffb0 offset: 01a0

>>>stack>>>
3fffff80:  40214cad 3fffdab0 00000002 3ffefdc0  
3fffff90:  402146bc 3fffdab0 00000000 3fffdcc0  
3fffffa0:  3ffe9900 40000f49 3fffdab0 40000f49  
<<<stack<<<

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x4010f000, len 1264, room 16 
tail 0
chksum 0x42
csum 0x42
~ld
..
WiFi connected
Server started
192.168.1.222

**Comment by eumb** [comment]
Same issue here.

Does anyone got a solution?

**Comment by Links2004** [comment]
with out a objdump or at least a call stack analyses its hart to say where the error is coming from.
 - install the ESP Exception Decoder tool https://github.com/me-no-dev/EspExceptionDecoder
 - flash the ESP new from the IDE.
 - wait for stack dump
 - c&p the error to analyses 
 - post the result here

![ESP Exception Decoder](https://files.gitter.im/esp8266/Arduino/Ebij/Screen-Shot-2016-01-28-at-11.29.34.png)

**Comment by Pablo2048** [comment]
so here is one example. Hope that help...
![snimek z 2016-02-20 17-26-34](https://cloud.githubusercontent.com/assets/6281647/13197760/3ea41f6a-d7f7-11e5-8176-bf3a5e4fce7d.png)

**Comment by Links2004** [comment]
can you try the same with latest git?
we have changed the heap management to umm.
since the SDK one make too much problems.

**Comment by Pablo2048** [comment]
I can try, but I'm very beginner with Linux and I'm very happy that Arduino with ESP8266 run, so it will take some time ;-)

**Comment by Pablo2048** [comment]
Ok, so I installed actual git version. I'm doing heavy testing right now (last error took me about 1/2 hour to generate...).

**Comment by Pablo2048** [comment]
so here is the result:
Uptime 0 days 00:56:06.628
Requests 1600 (2 seconds autorefresh)
It seems like this error is gone. Thank You Links2004.

**Comment by jeferrb** [comment]
If I'm sure, I'm getting the same problem as Pablo.
I'm using the last git version.
Have someone any advice?

There are three samples below (split by count.S:109):
![screenshot from 2016-03-03 00 18 19](https://cloud.githubusercontent.com/assets/12548000/13483721/cd75db98-e0d5-11e5-9358-c98ecb80ca48.png)

**Comment by igrr** [comment]
Hi @jeferrb, what makes you think the exception is related to WebServer? I don't see a single stack frame from ESP8266WebServer library in the screenshot you've posted.

**Comment by jeferrb** [comment]
Hi @igrr, thanks for reply. It is really not a WebServer question. Since it stops on the same line I thought it's related. I'm trying to port a project that works on Arduino Mega and I need some hint to understand where these error come from.

**Comment by igrr** [comment]
The final line in stack trace is actually the first function executed. Stack traces are normally read from bottom to top. I.e. the last function is actually `GetRandomBitRaw2`.
You might want to read https://github.com/esp8266/Arduino/blob/master/doc/exception_causes.md to get better understanding of exception line at the top of stack trace.

**Comment by jeferrb** [comment]
Hum, I get it now. So many thanks igrr, you help a lot. And, you are doing a grate job.

**Comment by bamboklaat** [comment]
Hi everyone, sort of late, but I found myself in a problem related to this one (I think). I started with arduino UNO since I found the loading much faster that ESP, so it was ok untill I started to migrate, my sistem is a LCD & rtc1307 (I2c) and a rotary encoder (with push bottom). My 1st step was to make the 3 periphericals work ok on the ESP, and it did work, just a few modification needed, the point was to make the whole code work ok, I got the EXEPTION (9) (Load or store to an unaligned address) when the interrupt take place. First thing I did was to remove the "lcd.createChar(0, grado); "  believing that the "address 0" was causing problems, but the problem remains, also tried a small script to check it out, and this character appeared on the screen, so it doesnt seems to be that. later on, tried the "TIME" pin in the 3 possible states, when low and high, the LCD stays frozen in the main manue, leaving the time pin floating, at least it interrupts, goes to the main manu (showing the current time from rtc) and then fades away as it should.

Unfortunally, with my knowledge I can not go further than what I tried, any suggestion? I paste the code in case you find a critic variable around. Thanks a lot for reading.

`define encoder0PinA 13
#define encoder0PinB 12
#define buttonPin 14

// Librerias necesarias
#include "Wire.h"
#include "LiquidCrystal_I2C.h"
#include "DS1307new.h"
#include <Wire.h>  // Comes with Arduino IDE
#include <LiquidCrystal_I2C.h>

LiquidCrystal_I2C lcd(0x27, 2, 1, 0, 4, 5, 6, 7, 3, POSITIVE);  // Set the LCD I2C address
byte grado[8] = {B11100, B10100, B11100, B00000, B00000, B00000, B00000, B00000};           //Simbolo °
byte termometro[8] = {B00100, B01010, B01010, B01110, B01110, B11111, B11111, B01110};      //icon for termometer
byte hume[8] = {B00100, B00100, B01010, B01010, B10001, B10001, B10001, B01110};
byte bar[8] = {B11111, B11111, B11111, B11111, B11111, B11111, B11111, B11111};

int t, h;

#define MON_BUF_LEN 128

String mes, dia, anno;
String hora, minuto, segundo;
String vacia, horaChar, fechaChar;
char horaC[10], fechaC[11];
int second_old, minute_old, hour_old, monthDay_old, month_old, year_old;

char mon_buf[MON_BUF_LEN];
uint8_t mon_buf_cnt = 0;
char *mon_curr;

uint8_t mon_is_date;
uint16_t mon_year;
uint8_t mon_month;
uint8_t mon_day;

uint8_t mon_is_time;
uint8_t mon_hour;
uint8_t mon_minute;
uint8_t mon_second;
uint8_t mon_error;

unsigned long currentMillis;
unsigned long inter_backlight = 8000;      // va hasta 32K = 32 seg
unsigned long m;

volatile int lastEncoded = 0;
volatile long encoderValue = 0;
bool CW = 0;
int encoded;

int vel_vent = 50;
int var_hora_ini = 6;
int var_Periodo = 18;

long lastencoderValue = 0;

int lastMSB = 0;
int lastLSB = 0;

int encoder = 11;     // 10 u 11, defino el arranque

int MenuBoxTop = 2;          // Number of items in the Send Menu update this  if you change the number of menu items
int cityBoxTop = 4;       // if you change the submenu you have the changes this value as well
int HoraBoxTop = 3;

int knobMax = MenuBoxTop + 10; // Use to cycle the menus around in a loop so that item first comes after item last
int knobMin = 10;

bool Pri = 1;
bool bandera = 0;
int buttonState;

// flags for the sub-menu items.
boolean HoraMode = false;
boolean hora_iniMode = false;
boolean PeriodoMode = false;

boolean VentMode = false;
boolean vent_flag = 0;

String MenuBox[] = {
  " Configurar Horarios",
  " Velocidad Ventilador"

};

char HoraBox[] = "1Hora inicio 2Periodo 3back 4";

void setup() {
  Serial.begin (9600);

  pinMode(buttonPin, INPUT_PULLUP);
  pinMode(encoder0PinA, INPUT_PULLUP);
  pinMode(encoder0PinB, INPUT_PULLUP);

  attachInterrupt(encoder0PinA, updateEncoder, CHANGE);

  lcd.begin(16, 2);
  lcd.clear();
  lcd.home();
  //lcd.createChar(0, grado);   //Setup  termometro  hume
  //lcd.createChar(1, termometro);
  //lcd.createChar(2, hume);
  //lcd.createChar(3, bar);

  printDate();
  currentMillis = millis();

}

// --------------------INT ENCODER----------------------
void updateEncoder() {
  bandera = 1;
  delay(100);
  int MSB = digitalRead(encoder0PinA); //MSB = most significant bit
  int LSB = digitalRead(encoder0PinB); //LSB = least significant bit
  encoded = (MSB << 1) | LSB; //converting the 2 pin value to single number
  int sum  = (lastEncoded << 2) | encoded; //adding it to the previous encoded value

  if (sum == 0b1101 || sum == 0b0100 || sum == 0b0010 || sum == 0b1011) {
    encoderValue ++;
    CW = 1;
  }
  if (sum == 0b1110 || sum == 0b0111 || sum == 0b0001 || sum == 0b1000) {
    encoderValue --;
    CW = 0;
  }
  lastEncoded = encoded; //store this value for next time, UTILIZADO PARA SUMAR O RESTAR (IMAGINO)
}



// --------------------printDate----------------------
void printDate() {
  lcd.clear();
  RTC.getTime();
  int second = RTC.second;
  int minute = RTC.minute;
  int hour = RTC.hour; //24 hour tim
  lcd.setCursor(0, 1);
  if (hour < 10) {
    //lcd.print("0");
    lcd.setCursor(1, 1);
    lcd.print(hour);
  }
  else {
    lcd.print(hour);
  }

  lcd.setCursor(2, 1);
  lcd.print(":");
  lcd.setCursor(3, 1);
  if (minute < 10) {
    lcd.print("0");
    lcd.setCursor(4, 1);
    lcd.print(minute);
  }
  else {
    lcd.print(minute);
  }
  lcd.setCursor(5, 1);
  lcd.print("hrs");
  //  h = dht.readHumidity();
  //  t = dht.readTemperature();

  lcd.setCursor(0, 0);
  //  lcd.write(byte(1));
  lcd.setCursor(2, 0);
  //  lcd.print(t);
  lcd.setCursor(4, 0);
  // lcd.write(byte(0));
  lcd.setCursor(5, 0);
  lcd.print("C");

  lcd.setCursor(9, 0);
  //  lcd.write(byte(2));
  lcd.setCursor(11, 0);
  //  lcd.print(h);
  lcd.setCursor(13, 0);
  lcd.print("%RH");

}
// --------------------RESET----------------------
void inicio_() {
  VentMode = false;
  HoraMode = false;
  hora_iniMode = false;
  PeriodoMode = false;
  knobMax = MenuBoxTop + 10;
  knobMin = 10;
  encoder = 11;
  printDate();
  vent_flag = 0;
  //  back_f = 0;
  Serial.println("reset");
}
////////////////////////////////////LOOP//////////////////////////////////////////////////////////
void loop()
{
  buttonState = digitalRead(buttonPin);
  delay(1);

  if (!buttonState) {
    //Serial.print("ccc");
    //delay(500);

    if (millis() > currentMillis + inter_backlight) {
      currentMillis = millis();
      lcd.noBacklight();
      inicio_();
      Pri = 1;
    }

  } else {
    delay(500);
    currentMillis = millis();
    lcd.backlight();

    if (Pri) {
      printDate();
      Pri = 0;
    }
    else {

      /*
          Serial.print("HoraMode ");
          Serial.println(HoraMode);
          Serial.print("PeriodoMode ");
          Serial.println(PeriodoMode);
          Serial.print("hora_iniMode ");
          Serial.println(hora_iniMode);
          Serial.print("VentMode ");
          Serial.println(VentMode);
      */
      // para que detecte uno solo push
      if (encoder == 10 && !HoraMode) {
        switchToHora ();
        return;
      } else if (encoder == 11 && !VentMode) {
        switchToVent();
        return;
      }
      if (encoder == 1 && !hora_iniMode) {
        Serial.println("Entro condicion enc=1 ");
        ShowHora_ini ();
        Serial.print("hora_iniMode ");
        Serial.println(hora_iniMode);
        return;
      } else if (encoder == 2 && !PeriodoMode) {
        ShowPeriodo ();
        return;
      } else if (encoder == 3 && HoraMode) {   //&& !HoraMode
        Serial.print("inicio__1_1_1eo entre al back");
        inicio_ ();
        return;
      }
      if (VentMode) {
        showVent();
      }
      else if (HoraMode && !hora_iniMode  && !PeriodoMode) {
        showHoraBox();
      }
      else if (hora_iniMode && HoraMode) {
        ShowHora_ini();
      }
      else if (PeriodoMode && HoraMode) {       // condicion etapa
        ShowPeriodo();
      }

      if (vent_flag == 1) {                      // condicion de vel configuracion hecha
        inicio_();
      }
    }
  }

  if (bandera) {
    Serial.print("BANDERA ");

    if (CW == 1) {
      delay(100);
      encoder ++;
    }
    else {
      encoder --;
    }
    if (encoder < knobMin) {
      encoder = knobMax - 1; // lower limit of any menu roll back to top
    }
    if (encoder > knobMax - 1) {
      encoder = knobMin; // upper limit of any menu roll back to bottom
    }

    Serial.print("encoder ");
    Serial.println(encoder);
    //lcd.print(encoder);
    Serial.print("encoder real: ");
    Serial.println(encoderValue);

    bandera = 0;
  }
}

///////////////////////////////////////////SHOW BOX/////////////////////////////////////////////////////////////////////////////////////////////////////
void showMenuBox() {
  knobMax = MenuBoxTop + 10; // set the value ready for showKeyBox
  knobMin = 10;
  int stopHere = MenuBox[encoder - 10].length();
  for (int i = 1; i < stopHere ; i++) {
    if (encoder == 10) {
      if (i > 10) {
        lcd.setCursor(i - 6, 1);
        lcd.print(MenuBox[encoder - 10].charAt(i));
      }
      else {
        lcd.setCursor(i - 1, 0);
        lcd.print(MenuBox[encoder - 10].charAt(i));
      }
    }
    if (encoder == 11) {
      if (i > 9) {
        lcd.setCursor(i - 6, 1);
        lcd.print(MenuBox[encoder - 10].charAt(i));
      }
      else {
        lcd.setCursor(i - 1, 0);
        lcd.print(MenuBox[encoder - 10].charAt(i));
      }
    }
  }
}
//////////////////////////////////////////////////////
// HoraBox contains the command menu in one long string
void showHoraBox() {
  int startString = encoder + 48; // get the ascii code of this item ascii 48 is equal to 0
  int stopString = encoder + 49; // get the ascii code of the next item
  int i = 0;
  while (HoraBox[i] != startString) {
    i++; // skip over everything until you get to our item number
  }
  while (HoraBox[i] != stopString) {         // now start printing until we get to the next item number
    i++;                                    // bump up one, also means we don't print the item number
    if (HoraBox[i] != stopString) {          // don't print the next item number either
      lcd.print(HoraBox[i]);                 // print the menu item letter by letter
    }
  }
}

//////////////////////////////////////////////////////
// CityBox contains the command menu in one long string
void showVent() {
  if (CW == 1) {
    vel_vent = vel_vent + 10;
    if (vel_vent > 100) {
      vel_vent = 0;
    }
  }
  else {
    vel_vent = vel_vent - 10;
    if (vel_vent < 0) {
      vel_vent = 100;
    }
  }
  lcd.clear();
  lcd.print("Vel vent ");
  lcd.setCursor(13, 1);
  lcd.print(vel_vent);


  lcd.setCursor(vel_vent / 10, 1);
  lcd.write(byte(3));

  VentMode = true;    // set the command mode flag -- we're no longer sending

  vent_flag = 1;
}

void ShowHora_ini() {

  if (CW == 1) {
    var_hora_ini++;
    if (var_hora_ini > 24) {
      var_hora_ini = 0;
    }
  }
  else {
    var_hora_ini--;
    if (var_hora_ini < 0) {
      var_hora_ini = 24;
    }
  }

  lcd.clear();
  lcd.print("Enciende:");
  lcd.setCursor(10, 1);
  lcd.print(var_hora_ini);
  lcd.setCursor(12, 1);
  lcd.print("hrs");
  hora_iniMode = true;    // set the command mode flag -- we're no longer sending
  knobMax = 12;       //Lo configuro para el menubox, en este caso en 10
  knobMin = 10;
  encoder = 10;
  vent_flag = 1;

}

void ShowPeriodo() {
  //  var_Periodo++;
  if (CW == 1) {
    var_Periodo++;
    if (var_Periodo > 24) {
      var_Periodo = 0;
    }
  }
  else {
    var_Periodo--;
    if (var_Periodo < 0) {
      var_Periodo = 24;
    }
  }

  lcd.clear();
  lcd.print("T.Enc.    T.Apag.");
  lcd.setCursor(1, 1);
  lcd.print(var_Periodo);
  lcd.setCursor(3, 1);
  lcd.print("hrs");


  lcd.setCursor(12, 1);
  lcd.print(24 - var_Periodo);
  lcd.setCursor(14, 1);
  lcd.print("hrs");

  PeriodoMode = true;    // set the command mode flag -- we're no longer sending
  knobMax = 12;       //Lo configuro para el menubox, en este caso en 10
  knobMin = 10;
  encoder = 10;
  vent_flag = 1;
}
/////////////////////////////////SWITCH////////////////////////////////////////////////////

void switchToHora() {    //switchToCar()

  lcd.clear();
  // lcd.print("Conf Horaria");
  //showHoraBox();

  HoraMode = true;    // set the command mode flag -- we're no longer sending
  knobMax = HoraBoxTop + 1; // set the limits for scrolling the menu
  knobMin = 1;
  encoder = 0;       // put us on the top item
  showHoraBox();
}

void switchToVent() {   //switchToCity

  lcd.clear();
  showVent();
  VentMode = true;    // set the command mode flag -- we're no longer sending
  knobMax = 12;
  knobMin = 11;
  encoder = 11;       // DEBE VOLVER CON ESE 11, SINO ENTRA EN EL MODO TIEMPO (10) ver loop clik botton


}

// Funciones del Reloj
void mon_skip_space(void)
{
  for (;;)
  {
    if ( *mon_curr == '\0' )
      break;
    if ( *mon_curr > ' ')
      break;
    mon_curr++;
  }
}

uint16_t mon_get_value(void)
{
  uint16_t v = 0;
  uint8_t c;
  for (;;)
  {
    if ( *mon_curr >= '0' && *mon_curr <= '9' )
    {
      c = *mon_curr;
      c -= '0';
      v *= 10;
      v += c;
      mon_curr++;
    }
    else
      break;
  }
  mon_skip_space();
  return v;
}

uint8_t mon_check(char c)
{
  if ( *mon_curr != c )
    return 0;
  mon_curr++;
  mon_skip_space();
  return 1;
}

/*
  yyyy-mm-dd
*/
uint8_t mon_get_date(void)
{
  mon_year = mon_get_value();
  if ( mon_year < 100 )
    mon_year += 2000;
  if ( mon_check('-') == 0 )
  {
    mon_error = 1;
    return 0;
  }
  mon_month = mon_get_value();
  if ( mon_month == 0 || mon_month > 12 )
  {
    mon_error = 3;
    return 0;
  }
  if ( mon_check('-') == 0 )
  {
    mon_error = 2;
    return 0;
  }
  mon_day = mon_get_value();
  if ( mon_day == 0 || mon_day > 31 )
  {
    mon_error = 4;
    return 0;
  }
  mon_is_date = 1;
  return 1;
}


uint8_t mon_get_time(void)
{
  mon_second = 0;
  mon_hour = mon_get_value();
  if ( mon_hour > 23 )
  {
    mon_error = 6;
    return 0;
  }
  if ( mon_check(':') == 0 )
  {
    mon_error = 5;
    return 0;
  }
  mon_minute = mon_get_value();
  if ( mon_minute > 59 )
  {
    mon_error = 7;
    return 0;
  }
  if ( mon_check(':') == 0 )
  {
    mon_is_time = 1;
    return 1;
  }
  mon_second = mon_get_value();
  if ( mon_second > 59 )
  {
    mon_error = 8;
    return 0;
  }
  mon_is_time = 1;
  return 1;
}


uint8_t mon_get_date_time(void)
{
  char *mon_ptr;
  uint16_t v;

  mon_is_date = 0;
  mon_year = 2000;
  mon_month = 1;
  mon_day = 1;

  mon_is_time = 0;
  mon_hour = 0;
  mon_minute = 0;
  mon_second = 0;

  for (;;)
  {
    mon_ptr = mon_curr;
    v = mon_get_value();
    if ( mon_check(':') != 0 )
    {
      mon_curr = mon_ptr;
      if ( mon_get_time() == 0 )
        return 0;
    }
    else if ( mon_check('-') != 0 )
    {
      mon_curr = mon_ptr;
      if ( mon_get_date() == 0 )
        return 0;
    }
    else
      break;
  }
  return 1;
}

void mon_print_error(void)
{
  Serial.print("Errorcode: ");
  Serial.println(mon_error, DEC);
}

void mon_print_2d(uint8_t v)
{
  if ( v < 10 )
    Serial.print("0");
  Serial.print(v, DEC);
}

void mon_print_3d(uint8_t v)
{
  if ( v < 10 )
    Serial.print(" ");
  if ( v < 100 )
    Serial.print(" ");
  Serial.print(v, DEC);
}

void mon_print_date(uint16_t y, uint8_t m, uint8_t d)
{
  Serial.print(y, DEC);
  Serial.print("-");
  mon_print_2d(m);
  Serial.print("-");
  mon_print_2d(d);
}

void mon_print_time(uint16_t h, uint8_t m, uint8_t s)
{
  mon_print_2d(h);
  Serial.print(":");
  mon_print_2d(m);
  Serial.print(":");
  mon_print_2d(s);
}

void mon_info(void)
{
  char wd[7][3] = { "So", "Mo", "Tu", "We", "Th", "Fr", "Sa" };
  RTC.getTime();
  mon_print_date(RTC.year, RTC.month, RTC.day);
  mon_print_time(RTC.hour, RTC.minute, RTC.second);

}

void mon_dst(void)
{
 
  uint8_t isSummerTime;
  uint16_t i;

  m = millis();
  for ( i = 0; i < 1000; i++ )
    isSummerTime = RTC.isMEZSummerTime();
  m = millis() - m;

  m = millis();
  for ( i = 0; i < 1000; i++ )
    RTC.fillByTime2000(RTC.time2000);     // speed measue
  m = millis() - m;
  Serial.print(m, DEC);
  Serial.println("us)");

  m = millis();
  RTC.fillByYMD(RTC.year, 4, 1);    // first of April
  if ( RTC.dow == 0 )
    RTC.fillByCDN(RTC.cdn - 7);     // sunday before
  else
    RTC.fillByCDN(RTC.cdn - RTC.dow); // sunday before
  m = millis() - m;

  mon_print_date(RTC.year, RTC.month, RTC.day);


  RTC.fillByYMD(RTC.year, 11, 1);   // first of november
  if ( RTC.dow == 0 )
    RTC.fillByCDN(RTC.cdn - 7);     // sunday before
  else
    RTC.fillByCDN(RTC.cdn - RTC.dow); // sunday before

  mon_print_date(RTC.year, RTC.month, RTC.day);

}

void mon_set_date_time(void)
{
  if ( mon_get_date_time() == 0 )
  {
    mon_print_error();
    return;
  }
  RTC.getTime();
  //  Serial.print("Assign ");
  if ( mon_is_date != 0 )
  {
    mon_print_date(mon_year, mon_month, mon_day);
    //  Serial.print(" ");
    RTC.fillByYMD(mon_year, mon_month, mon_day);
  }
  if ( mon_is_time != 0 )
  {
    mon_print_time(mon_hour, mon_minute, mon_second);
    //   Serial.print(" ");
    RTC.fillByHMS(mon_hour, mon_minute, mon_second);
  }
  RTC.setTime();
  RTC.startClock();
}

void mon_list(void)
{
  uint8_t i, addr;
  uint8_t ram[8];

  for ( addr = 0; addr < 56; addr += 8 )
  {
    RTC.getRAM(addr, ram, 8);
    mon_print_3d(addr);
    Serial.print(": ");
    for ( i = 0; i < 8; i++ )
    {
      mon_print_3d(ram[i]);
    }
  }
}

void mem_memory(void)
{
  uint8_t a, m;
  a = mon_get_value();
  m = mon_get_value();

}

void exec(void)
{
  //  Serial.println(mon_buf);
  mon_curr = mon_buf;

}`

**Comment by MHz000** [comment]
as a newbie I've the same problem and **no idea how to proceed  :-((**  

Exception (9):
epc1=0x40202846 epc2=0x00000000 epc3=0x00000000 excvaddr=0x3ffe8542 depc=0x00000000

ctx: cont
sp: 3ffefa00 end: 3ffefc10 offset: 01a0

>>>stack>>>
3ffefba0:  3ffeeab0 3ffee4b0 0000003a 40202866 
3ffefbb0:  3fffdad0 3ffee4b0 3ffeebb8 40201da5 
3ffefbc0:  00000000 00000000 00000000 40202c78 
3ffefbd0:  00000000 feefeffe 3ffeebb8 40202c9c 
3ffefbe0:  00000000 00000000 00000001 3ffeebe4 
3ffefbf0:  3fffdad0 00000000 3ffeebdc 40203150 
3ffefc00:  feefeffe feefeffe 3ffeebf0 40100718 
<<<stack<<<

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x4010f000, len 1384, room 16
tail 8
chksum 0x2d
csum 0x2d
v09f0c112
~ld

after one or two successfull requests.

It woud be fine if someone could make a little test on his system in his envireoment to see it something is wrong in my setup.
Here is the simplified code
//--------------------------------------------------------------------------------------------------------------------------------------------------------------------
char versions_id[] = "ESP__WEB_test_V6";
#include <ESP8266WiFi.h>
#include <WiFiClient.h>
#include <ArduinoJson.h>
#include <TimeLib.h>

const char WIFI_SSID[] = "...";               // WiFi information
const char WIFI_PSK[] = "...";

const char http_site[] = "api.usno.navy.mil";       // Remote site information
const int http_port = 80;

const int LED_PIN = 16;                                   // global variables & definitions
WiFiClient client;
char c;                                                              // incomming char from WEB-server
unsigned long delayMillis;                               // delay loop time
unsigned long pause = 20000;                        // prog paused

// JSON stuff
StaticJsonBuffer<1000> jsonBuffer;                  // Memory pool for JSON object tree.
char json[500];                                     // JSON object char array, filld with server response
String head = "";                                   // HTTP header from server
int json_i=0;                                       // index for json char array
int JS_bracket = 0;                                 // counter for JSON { and } brackets 

void setup() {
  Serial.begin(115200);                             // Set up serial console to read web page
  Serial.println(); Serial.println(versions_id);
  Serial.println("********************RE-/START******************");
  pinMode(LED_PIN, OUTPUT);                         // Set up LED for debugging
}

void loop() {
  head = "";                                        // reset server response string
  JS_bracket = 0;                                   // reset json bracket counter
 
  connectWiFi();                                    // Connect to WiFi
 
  if ( !getPage() ) {                               // Attempt to connect to website   
    Serial.println("GET request failed");}
    else {
    Serial.println("GET request ok");}
   
    while (client.connected() ) {
    if ( client.available() ) {                     // If there are incoming bytes, print them
      c = client.read();
      //Serial.print(c);
      if (JS_bracket == 0) {head=head+c;}           // fill in server response
      if (c == '{') {++JS_bracket;}                 // ectraxt json object
      if (c == '}') {--JS_bracket; }
      // still to do: skip white space outside ".."
      if (JS_bracket > 0) { //Serial.print(c);      // look for matching closing bracket
      json[json_i]=c; json_i++;}; }
    }
    json[json_i]='}';                               // add a closing } bracket 
    Serial.println("stop client");                  // If the server has disconnected, stop the client and WiFi
    client.stop();                                  // Close socket and wait for disconnect from WiFi
    if ( WiFi.status() != WL_DISCONNECTED ) {
      WiFi.disconnect();}
    digitalWrite(LED_PIN, HIGH);                    // Turn off LED
    
// *****process server data here
  Serial.println("processing server data");
//  for (int i=0; i<=json_i; i++) {Serial.print(json[i]);}  Serial.println();  // control print
  Serial.println("******************pausing****************");
  delayMillis= millis()+pause;                             // 5 minutes intervall 
   do {
     yield();
     } while (millis() < delayMillis);    
  Serial.println("***********start after pausing***********");
// *****end processing server data  
  
}

//******************connect WiFi***********************
void connectWiFi() {                            // connect WiFi
byte led_status = 0;
WiFi.mode(WIFI_STA);                            // Set WiFi mode to station (client)
WiFi.begin(WIFI_SSID, WIFI_PSK);                // Initiate connection with SSID and PSK
while ( WiFi.status() != WL_CONNECTED ) {       // Blink LED while we wait for WiFi connection
  yield();
  digitalWrite(LED_PIN, led_status);
  led_status ^= 0x01;
  delay(100);
  }
digitalWrite(LED_PIN, HIGH);                    // Turn LED on when we are connected
}

//***********************get page***************
bool getPage() {                                // Perform an HTTP GET request to a remote page
if ( !client.connect(http_site, http_port) ) {  // Attempt to make a connection to the remote server
return false;
}
client.println("GET /rstt/oneday?date=9/14/2016&coords=47.73N,9.16E&tz=1 HTTP/1.1"); // Make an HTTP GET request
client.print("Host: ");
client.println(http_site);
client.println("Connection: close");
client.println();
return true;
}
//------------------------------------------------------------------------------------------------------------------------
enviroment:  WIN10 arduion 1.6.12

Thank you for any help or hint

**Comment by borqborq** [comment]
I had the same problem: Exception (9). My issue was that the page would render for the client, but you couldn't submit the tiny form that was on that page. It threw the error every time. 

A 'tremor in the force' felt like it was memory related - I thought maybe the array capturing the clients HTTP response was too big and popping the RAM so I kept trimming the array variable allocation down with no love...

On a whim, I made the array size bigger and the error vanished. The 8266 web server runs smooth 99.9% (because nothing is ever 100% in IOT, right?)

It turned out that the array that was capturing the HTTP_RESPONSE with was not big enough (by a hair). I originally allocated an array size of 500, which was enough to allow the web client to load or refresh the page successfully, but if there was any other data added from the client and sent to the 8266 (i.e. a tiny web form on the page), it would throw the error after submitting the form.

`char http_resp[500] = {0};`

After changing the array size to 700, the error was gone, and the 8266 takes the form submission and processes it just fine.

`char http_resp[700] = {0};`

Long story short, my problem was in the handling of the arrays and HTTP responses from the client.

Hope this helps someone out there!

**Comment by rajsingha** [comment]
> ah I see. thanks @igrr! I'll alter the code not to close the connection each time and see if that fixes the crashing.

Please tell me how did you fixed it?

**Comment by johny77g** [comment]
Hi, had similar problem with exception 9. Finally in BoardsManager in esp8266 I installed older version 2.3.0 - additionally I had to choose Board Wemos D1 retired - and problem with exception 9 disappeared completely.

**Comment by riskiadi** [comment]
Exception (9):
20:33:41.692 -> epc1=0x4020b648 epc2=0x00000000 epc3=0x40000f68 excvaddr=0x00000041 depc=0x00000000

after add firebase database library and arduinojson version 5 latest

**Comment by devyte** [comment]
I understand that firebase uses WiFiClientSecure, research that and you'll know what's wrong.

**Comment by bappapan220** [comment]
Thank you Sir. thanks a lot.
